ms2pip/ms2pip_tools/spectrum_output.py

```diff
@@ -3,492 +3,678 @@
+                last_line = line
+            last_scannr = int(last_line.split("\t")[1])
+        return last_scannr
+
+    def _generate_ssl_modification_mapping(self):
+        """
+        Make modification name -> ssl modification name mapping.
+        """
+        self.ssl_modification_mapping = {
+            ptm.split(",")[0]: "{:+.1f}".format(round(float(ptm.split(",")[1]), 1))
+            for ptm in self.params["ptm"]
+        }
+
+    def _get_ssl_modified_sequence(self, sequence, modifications):
+        """
+        Build BiblioSpec SSL modified sequence string.
+        """
+        pep = list(sequence)
+
+        for loc, name in zip(
+            modifications.split("|")[::2], modifications.split("|")[1::2]
+        ):
+            # C-term mod
+            if loc == "-1":
+                pep[-1] = pep[-1] + "[{}]".format(self.ssl_modification_mapping[name])
+            # N-term mod
+            elif loc == "0":
+                pep[0] = pep[0] + "[{}]".format(self.ssl_modification_mapping[name])
+            # Normal mod
+            else:
+                pep[int(loc) - 1] = pep[int(loc) - 1] + "[{}]".format(
+                    self.ssl_modification_mapping[name]
+                )
+        return "".join(pep)
+
+    def _write_bibliospec_core(self, file_obj_ssl, file_obj_ms2, start_scannr=0):
+        """
+        Construct Bibliospec SSL/MS2 strings and write to file_objects.
+        """
+
+        for i, spec_id in enumerate(sorted(self.preds_dict.keys())):
+            scannr = i + start_scannr
+            seq = self.peprec_dict[spec_id]["peptide"]
+            mods = self.peprec_dict[spec_id]["modifications"]
+            charge = self.peprec_dict[spec_id]["charge"]
+            prec_mass, prec_mz = self._get_precursor_mz(seq, mods, charge)
+            ms2_filename = os.path.basename(self.output_filename) + "_predictions.ms2"
+
+            peaks = self._get_peak_string(
+                self.preds_dict[spec_id]["peaks"], sep="\t", include_annotations=False,
+            )
+
+            if isinstance(mods, str):
+                if mods != "-" and mods != "":
+                    mod_seq = self._get_ssl_modified_sequence(seq, mods)
+                else:
+                    mod_seq = seq
+            else:
+                mod_seq = seq
+
+            rt = self.peprec_dict[spec_id]["rt"] if self.has_rt else ""
+
+            file_obj_ssl.write(
+                "\t".join(
+                    [ms2_filename, str(scannr), str(charge), mod_seq, "", "", str(rt)]
+                )
+                + "\n"
+            )
+            file_obj_ms2.write(
+                "\n".join(
+                    [
+                        f"S\t{scannr}\t{prec_mz}",
+                        f"Z\t{charge}\t{prec_mass}",
+                        f"D\tseq\t{seq}",
+                        f"D\tmodified seq\t{mod_seq}",
+                        peaks,
+                    ]
+                )
+                + "\n"
+            )
```

Comment:
Use `writelines` for `file_obj_ms2`? And `file_obj_ssl` seems a normal tsv, so `csv` could be used instead.

Reply:
Added a TODO for csv